src/js/easymde.js

```diff
@@ -2058,6 +2085,42 @@ EasyMDE.prototype.uploadImage = function (file, onSuccess, onError) {
 
 };
 
+/**
+ * Upload an image to the server using a custom upload function.
+ *
+ * @param imageUploadFunction {Function} The custom function to upload the image passed in options
+ * @param file {File} The image to upload, as a HTML5 File object (https://developer.mozilla.org/en-US/docs/Web/API/File).
+ */
+EasyMDE.prototype.uploadImageUsingCustomFunction = function(imageUploadFunction, file) {
+    var self = this;
+    function onSuccess(imageUrl) {
+        afterImageUploaded(self, imageUrl);
+    }
+
+    function onError(errorMessage) {
+        var filledErrorMessage = fillErrorMessage(errorMessage);
+        // show error on status bar and reset after 10000ms
+        self.updateStatusBar('upload-image', filledErrorMessage);
+
+        setTimeout(function () {
+            self.updateStatusBar('upload-image', self.options.imageTexts.sbInit);
+        }, 10000);
+
+        // run error handler from options, this alerts the message.
+        self.options.errorCallback(filledErrorMessage);
+    }
+
+    function fillErrorMessage(errorMessage) {
+        var units = self.options.imageTexts.sizeUnits.split(',');
+        return errorMessage
+            .replace('#image_name#', file.name)
+            .replace('#image_size#', humanFileSize(file.size, units))
+            .replace('#image_max_size#', humanFileSize(self.options.imageMaxSize, units));
+    }
+
+    imageUploadFunction(file, onSuccess, onError);
```

Comment:
Maybe we can change this a little to keep the API simple.

```suggestion
    try {
        var url = imageUploadFunction(file);

    } catch(error) {
        if (typeof error === 'string') {
            onError(error);
        } else if (error instanceof Error) {
            onError(error.message);
        } else {
            console.error(error);
        }
        return;
    }

    onSuccess(url);
```

This way the user does not need to call the onSuccess and onError functions himself, just return a url, and any exceptions will be handled.
Making use of `return` and `throw` comes more naturally to developers than `onSuccess` and `onError` I think.

```
imageUploadFunction?: (file: File) => string;
```

What do you think?

```
var editor = new EasyMDE({
    uploadImage: true,
    imageUploadFunction: function(file) {
        // Thrown error can be handled.
        var imageUrl = uploadImage(file);
        return imageUrl;
    },
});
```

```
var editor = new EasyMDE({
    uploadImage: true,
    imageUploadFunction: function(file) {
        try {
            return uploadImage(file);
        } catch (error) {
            if (error.message === 'someSpecificError') {
          		throw new Error('Failed because of a reason');
            } else {
				throw error;
			}
            console.log(error);
        }
    },
});
```

Reply:
I totally agree and that was my first approach but to do it in that way, `uploadImage(file)` should return a `Promise<string>` because if for example, you want to upload the image to `S3` or call a Rest API with `axios`, that is asynchronous and we have to work with promises. I tried it and I had to update the target to `es2015` in tsconfig but some libraries started to fail. That's why I rolled back to callback functions.

Reply:
You're right, I did not consider the async nature of `uploadImage`, my bad.

Reply:
Something something TS/es6 rewrite :P

Reply:
hehe for sure!